### PR TITLE
[E2E] Completely rewrite and reinforce a `visitDashboard` helper

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### Status
PENDING REVIEW

### What does this PR accomplish?
- Reinforces the existing `visitDashboard` helper
    - It will automatically wait either for the `GET /api/dashboard/:id` or for the individual dashboard card queries
    - This alone solves a whole series of flakes with visualization options, multi-series cards, etc. where FE is failing if Cypress tries to click on any of the related icons while the dashboard card contents is still loading
- Adds e2e tests for this helper for each of the user roles
    - NOTE: These tests will most probably run on `master` in the future. For now, we're not running them in CI at all. They were meant as a sanity check for local testing.

On the surface, this helper seems simple - it just takes a dashboard id and visits that dashboard. But in reality, it is intelligently setting up a lot of `intercept`s and `wait`s based on the contents of the dashboard (or the lack of it).

It takes into account:
- empty dashboards
- dashboards with markdown cards only
- dashboards with GUI questions
- dashboards with native questions
- dashboards with models
- dashboards with pivot tables
- dashboards with two or more cards, with any combination of the above card types
- dashboards that a user doesn't even have permission to access
- dashboards that contain cards a user doesn't have a permission to view

### Explanation
We need to fire this request in order to obtain the needed info about the dashboard. Because some users don't have permissions to issue this request, there's a nice escape hatch in Cypress called `failOnStatusCode` which we can set to false. It has an effect on this and only this request. The rest of the helper and all related tests will still fail on exceptions.

```js
  cy.request({
    method: "GET",
    url: `/api/dashboard/${dashboard_id}`,
    failOnStatusCode: false,
  })
```
We obtain `ordered_cards` (all dashboard cards) from the previous step and from that point onward, it's just a matter of proper filtering.

For all valid questions, we set an intercept, create an alias for it and then wait for all aliases:
`cy.wait(["@alias1", "@alias2", "@aliasN"]);`

<details>
  <summary>Click to expand and to see the screenshot of all passed tests</summary>

image

</details>

### CODE REVIEW

**Comment by nemanjaglumac** [comment]
@ariya after this PR, it **should be** safe to move `dashboard` and `dashboard-filters` groups to E2E on PR workflow.

File: frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js [hunk 1]
```diff
@@ -133,28 +133,80 @@ export function visitQuestion(id) {
...
+function hasAccess(statusCode) {
   return statusCode !== 403;
 }
+
+function dashboardHasQuestions(cards) {
+  if (Array.isArray(cards) && cards.length > 0) {
+    const questions = cards
+      // Filter out markdown cards
+      .filter(({ card_id }) => {
+        return card_id !== null;
```
Comment (ariya) [inline frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js:199]: Nitpick: could use the short form of arrow function (no need for explicit `return`).

File: frontend/test/__support__/e2e/integration/visit-dashboard.cy.spec.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  Object.keys(Cypress._.omit(USERS, "sandboxed")).forEach(user => {
+    context(`${user.toUpperCase()}`, () => {
+      beforeEach(() => {
+        restore();
+        cy.signInAsAdmin();
+
+        setup();
+
+        if (user !== "admin") {
+          cy.signIn(user);
```
Comment (ariya) [inline frontend/test/__support__/e2e/integration/visit-dashboard.cy.spec.js:16]: BTW what happens if `admin` user is going through `cy.signIn`?

Comment (ariya) [inline frontend/test/__support__/e2e/integration/visit-dashboard.cy.spec.js:16]: BTW what happens if `admin` user is going through `cy.signIn`?

File: frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js [hunk 3]
```diff
@@ -133,28 +133,80 @@ export function visitQuestion(id) {
...
-    canViewDashboard(statusCode) && cy.wait("@getParentCollection");
+export function visitDashboard(dashboard_id) {
+  // Some users will not have permissions for this request
+  cy.request({
+    method: "GET",
+    url: `/api/dashboard/${dashboard_id}`,
+    // That's why we have to ignore failures
+    failOnStatusCode: false,
+  }).then(({ status, body: { ordered_cards } }) => {
+    const dashboardAlias = "getDashboard" + dashboard_id;
```
Comment (ariya) [inline frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js:148]: Nitpick: `getDashboard ${dashboard_id}`

(if we aim for consistency with the following line)

Comment (ariya) [inline frontend/test/__support__/e2e/helpers/e2e-misc-helpers.js:148]: Nitpick: `getDashboard ${dashboard_id}`

(if we aim for consistency with the following line)

**Review by ariya — APPROVED** [review]
Let's ship this :medal_sports:

File: frontend/test/__support__/e2e/integration/visit-dashboard.cy.spec.js (hunk 2 above)
Comment (nemanjaglumac) [reply frontend/test/__support__/e2e/integration/visit-dashboard.cy.spec.js:16]: https://github.com/metabase/metabase/pull/21801/files#diff-248e295e41a02633b63797eaff18f1d1f86ff8e82bdf82ac5a0d542846b0bc4cR11

We're already signed in as admin. No need to do it twice.
And it's crucial to sign in as admin in order to do the `setup`.

**Pull request merged**